Add segmentation comparison features (model grouping, layout control, structure-wise review) - Auto layout#2

LenaGiebeler opened the pull request:
Related to Issue [#2 ](https://github.com/cpinter/SlicerSegmentationVerification/issues/2)

**Demo Video:**

https://github.com/user-attachments/assets/8fbc35e4-bc97-4c9a-8a95-fd126b988b28

**Current Status:**
- Segmentations can be selected and reviewed based on the chosen volume.
- Supports visualization of segmentations and models in both 3D and 2D views (axial, sagittal, coronal).
- Segmentation models group multiple related segmentations via keyword-matching; customizable via a pop-up table.
- Layout configuration: horizontal (default) or vertical arrangement of 2D/3D views.
- **Collapsible Options panel** allows: Linking/unlinking of 3D and 2D views; Switching between outline and solid segmentation display
- **Segment-by-segment comparison:** Navigate through individual segments in all views; Show neighboring structures in semi-transparent mode; Select segment groups (e.g., "Ribs"), which group segments based on name-based keywords. These groups can be edited via a configuration table, and when selected, only the segments in the chosen group are displayed in the views.

LenaGiebeler committed:
Possibility to generate x 3D views + corresponding 2D views easily

LenaGiebeler committed:
Resolved Problem that I can't reuse ID to set the layout

LenaGiebeler committed:
Load 3D View and 2D slices for each Model automatically. Link Views and Toogle 3D View Slcie Visibility

LenaGiebeler committed:
Loading and chaning views without clicking the apply Button. Changing between a horizontal and a vertical layout. Showing only the outlines of segmentations in 2D view

LenaGiebeler committed:
On Element Secetion in the structure Tabel only the structure is shown for each segmentation model in the corresponding view

LenaGiebeler committed:
Added Code to save and reload the scene

LenaGiebeler committed:
Add pop-up tables with grouping keywords for models and segmentation groups

LenaGiebeler committed:
Minor bug fix segment table selection; Code documentation

LenaGiebeler committed:
Update README.md

LenaGiebeler committed:
Update README.md

LenaGiebeler committed:
Update README.md

LenaGiebeler committed:
Rebase README

fedorov commented:
@LenaGiebeler whenever this is ready for testing - can you please add a comment and also include a data sample that can be used for testing?

LenaGiebeler commented:
@fedorov This is ready for testing now.
Data: https://www.dropbox.com/scl/fo/743hmzfb7qto3ajlqiex5/ANxrX0FZjrH8EK93rDH8I9E?rlkey=vmguq331kirjpwx8b4zmbfdsc&st=ich5aiuz&dl=0

fedorov commented:
@LenaGiebeler quick comment: `.DS_store` and `__pycache__` should be added to `.gitignore` and removed from the repo!

fedorov commented:
Overall, looks great for the initial review! Worked for me right away!

Comments from the user perspective - before I look at the code at all!
* I suggest disabling 3D view by default, since not all of the users will need it, but at the same time it can take significant time to create 3d renderings
* suggest enabling linking 2d views by default (this will also be consistent with 3d views, which are linked by default)
* "Segmentation comparison" "Select segmentations" dropdown is not updated when scene changes - new segmentations loaded into the scene do not show up, and removing segmentations from the scene does not remove items from the dropdown
* "Segment by Segment Comparison" table 
  * has dark background and black font color, even though my system theme is light - ideally this table should respond to the system theme
  * would it be possible to enable sorting of this table by clicking on the column headers? I think it would be good to know which of the segments are present in most segmentations
* I think it would be helpful to have either a tooltip or "?" button next to the "Select segmentation models" and "Select segmentation group" "+" buttons (or alternatively instructions on what to do can be included in the popup window), since I think it is no obvious at all how to use them
* nitpick: capitalization should be consistent for all labels - right now in "Show neighboring segments [...]" only first letter is capitalized, which is different from other labels (also "Change Segmentation representation [...]")
* Instead of "instantiate vertical layout" checkbox I think it is better to use "vertical"/"horizontal" radiobox
* switching from segmentation comparison to segmentation verification mode should reset the layout to single segmentation/volume, no comparison

Also, the functionality to load all segmentations that reference a given image series would be really handy here! I think it would be nice if you could add a button somewhere to do that! What do you think?

I think this set of suggestions is a good place for me to take a pause and pass it over to you @LenaGiebeler!

cpinter commented:
I watched the video, but I have a hard time understanding what is going on. The UI is incredibly complex, and the workflow seems convoluted as well. Should not this new part be in a separate module?

Some specific comments (to begin with):
* Remove the binary files from the branch, and amend the commit (we don't want those in the history either)
* As @fedorov says, capitalization is off. However, while he mentions that "Show neighboring segments" is the different one, I'd venture that all the others are different from the majority from Slicer. Sentence case is the style used in Slicer, not title case
* I find the text in the user interface quite hard to understand, for example "instantiate vertical layout" (instantiate is a programming term, the text should say what happens in the application not in the code) or "Change Segmentation representation..." (same here, it should be "Show outlines in 2D" or similar). Less technical jargon, more simple terms
* Make the layout non-adjustable, the columns fall apart (`split="false"`)
* I appreciate that you have your own coding style, but you added things to an existing code. Please adhere to that existing style.
   * And please do not mix styles (`_load_structures_in_table` vs `onSegmentSelectionChangedMultiple`, `"""` vs `'''`)
* Do not abbreviate things that much (`i`, `dn`, `seg`, `lm`, .....). Readability is way more important than the length of certain names, and this is just not done in the Slicer ecosystem for the same reason (even file names, there is no reason to abbreviate the word segmentation in `SegModelsDialog.ui`)
* Where do the "segmentation groups" come from? How are they defined?

fedorov commented:
@cpinter thank you for the initial review, we appreciate your time.

The idea to separate this functionality into an independent module is a good one. We will discuss it with Lena. 

Note that Lena is a student from Germany, visiting my lab just for a few month (leaving at the end of this month), with just a BS degree and not a lot of experience in software engineering or collaborative open source development. This is her first experience with developing in 3D Slicer.

If you feel this functionality does not add value to your extension, or the code style is not acceptable, or it is too much work to decide, we can contribute it in a separate extension. Please let us know.

fedorov commented:
> Should not this new part be in a separate module?

@cpinter would you like that module to be part of your extension, or should we put it into a separate extension?

LenaGiebeler committed:
#3 put Segmentation Comparison in its own module

LenaGiebeler committed:
Pull request #3:
- Disable 3D View by default; enable 2D View by defalt
- Enable 2D Link by default
- Update Gui, when new segmentations are loaded/deleted
- Segment table: fix colors, sort colimn feature
- Tooltip for "+" Button for segmentation models and segmentation groups
- vertical/horizontal radiobox
- Load segmentations for selected volume

fedorov commented:
More comments based on my testing today:
* I think "Add segmentation" can be confusing - how about "Load segmentations" instead? I would also put that button below the volume selector, and also next to the button show something like "X DICOM SEG series referencing this volume found" so the user has the idea what exactly will be loaded
* something seems to be off with the module panel layout  ![image](https://github.com/user-attachments/assets/fde379e8-57f9-4517-8a45-389125094759)
* for the pattern matching of the segmentations models, I think it would be helpful to have a button that gives explanation how this works, and also I think it would make sense to show the list of all series descriptions where pattern match will be applied, maybe also add a column with the descriptions that match the pattern as the third column in this table in the popup? ![image](https://github.com/user-attachments/assets/b27c9532-32eb-4c06-b5da-fe1d87148a23) - I later noticed you have this in a tooltip, not sure this is sufficient!
* "Show neighboring segments" does not seem to work for me (it seems to work for some structures but not for all)
* for the structures table, I would use "Segment name" and "Models containing segments" (or "Models") as column names
* at some point it seems like the application becomes not very responsive, 3d view takes long to update
* show neighbor segments does not apply immediately but only on the next segment selection

cpinter commented:
Sorry for the silence, we have an uncharacteristic deadline (still, next week, but then I disappear for two more weeks).

> not a lot of experience in software engineering or collaborative open source development

Noted! Then let's adjust the expectations. If the major things in my previous comments are fixed then it's a big step forward.

I am OK with integrating this module in the same extension, but I wanted to understand how it worked exactly, and I failed. If you @fedorov say that it works as you imagined, then I'll just accept it (will only have measurable time for this in August...).

fedorov commented:
Csaba, no worries, I thought you are very busy!

I recommended Lena to publish this as a separate extension - to avoid the confusion, but also so that she feels ownership of her work. I think it will be better this way. She is now working on some coursework and a written report, but she plans to finish this in the coming 1-2 months and publish the extension. There was quite some interest in her work at the last PW, so that was quite encouraging!

LenaGiebeler committed:
Bug fix - Show neighbouring semi transparent

LenaGiebeler committed:
Update README.md

fedorov commented:
This PR is superseded by the repository in https://github.com/ImagingDataCommons/CrossSegmentationExplorer. We will keep this functionality separate from this extension, at least for now
